File: apps/javascript/worker/src/ingestion/processors/markdown-header-chunker.ts
import { Injectable, Logger } from '@nestjs/common';
import { DocumentChunk, ChunkMetadata } from "../../models/document-chunk";
import { ChunkingProvider } from './chunking.provider';

/**
 * [마크다운 헤더 기반 청커]
 */
@Injectable()
export class MarkdownHeaderChunker extends ChunkingProvider {
    private readonly logger = new Logger(MarkdownHeaderChunker.name);

    splitDocument(
        rawText: string,
        commonMeta: Omit<ChunkMetadata, 'chunkIndex' | 'sectionTitle'>
    ): DocumentChunk[] {
        const lines = rawText.split('\n');
        const chunks: DocumentChunk[] = [];

        let currentSectionTitle = 'Root';
        let currentContent: string[] = [];
        let chunkIndex = 0;

        for (const line of lines) {
            if (line.startsWith('## ') || line.startsWith('### ')) {
                if (currentContent.length > 0) {
                    this.addChunk(chunks, currentContent, currentSectionTitle, chunkIndex++, commonMeta);
                    currentContent = [];
                }
                currentSectionTitle = line.replace(/#/g, '').trim();
            }
            currentContent.push(line);
        }

        if (currentContent.length > 0) {
            this.addChunk(chunks, currentContent, currentSectionTitle, chunkIndex++, commonMeta);
        }

        return chunks;
    }

    private addChunk(
        chunks: DocumentChunk[],
        content: string[],
        sectionTitle: string,
        chunkIndex: number,
        commonMeta: Omit<ChunkMetadata, 'chunkIndex' | 'sectionTitle'>
    ) {
        const text = content.join('\n').trim();
        if (!text) return;

        const id = `${commonMeta.filePath}#${sectionTitle}#${chunkIndex}`;

        chunks.push({
            id,
            text,
            metadata: {
                ...commonMeta,
                sectionTitle,
                chunkIndex
            }
        });
    }
}

File: backend/java/src/main/java/com/ragassistant/provider/chunking/MarkdownHeaderChunker.java
package com.ragassistant.provider.chunking;

import com.ragassistant.model.DocumentChunk;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class MarkdownHeaderChunker implements ChunkingProvider {

    @Override
    public List<DocumentChunk> splitDocument(String rawText, DocumentChunk.ChunkMetadata commonMeta) {
        String[] lines = rawText.split("\n");
        List<DocumentChunk> chunks = new ArrayList<>();

        String currentSectionTitle = "Root";
        List<String> currentContent = new ArrayList<>();
        int chunkIndex = 0;

        for (String line : lines) {
            if (line.startsWith("## ") || line.startsWith("### ")) {
                if (!currentContent.isEmpty()) {
                    addChunk(chunks, currentContent, currentSectionTitle, chunkIndex++, commonMeta);
                    currentContent = new ArrayList<>();
                }
                currentSectionTitle = line.replace("#", "").trim();
            }
            currentContent.add(line);
        }

        if (!currentContent.isEmpty()) {
            addChunk(chunks, currentContent, currentSectionTitle, chunkIndex++, commonMeta);
        }

        return chunks;
    }

    private void addChunk(List<DocumentChunk> chunks, List<String> content, String sectionTitle, int chunkIndex,
            DocumentChunk.ChunkMetadata commonMeta) {
        String text = String.join("\n", content).trim();
        if (text.isEmpty())
            return;

        String id = String.format("%s#%s#%d", commonMeta.getFilePath(), sectionTitle, chunkIndex);

        // Create a copy of commonMeta
        DocumentChunk.ChunkMetadata meta = DocumentChunk.ChunkMetadata.builder()
                .title(commonMeta.getTitle())
                .filePath(commonMeta.getFilePath())
                .docType(commonMeta.getDocType())
                .updatedAt(commonMeta.getUpdatedAt())
                .version(commonMeta.getVersion())
                .summary(commonMeta.getSummary())
                .permission(commonMeta.getPermission())
                .sectionTitle(sectionTitle)
                .chunkIndex(chunkIndex)
                .build();

        chunks.add(DocumentChunk.builder()
                .id(id)
                .text(text)
                .metadata(meta)
                .build());
    }
}
